### PR TITLE
Introduce causal mask

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: Introduce causal mask

This PR introduces a causal mask option is_causal, since current custom kernels do not support arbitrary masks.
(Also, building these masks is expensive, and avoiding to build them is a significant opportunity since causal mask is extremely common for training Transformer models.)

Test Plan: sandcastle & github ci/cd

Differential Revision: D41723137

### CODE REVIEW

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -278,8 +278,21 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
...
                 src_key_padding_mask_for_layers = None
 
+        if mask is not None:
+            if _causal_mask:
+                raise RuntimeError("specify either mask or _causal_mask, but not both")
+
+            sz = mask.size(0)
+            causal_comparison = torch.triu(
+                torch.ones(sz, sz, device=mask.device) * float('-inf'), diagonal=1
+            ).to(mask.dtype)
```
Comment (danthe3rd) [inline torch/nn/modules/transformer.py:288]: this might be very expensive, and will run every time the user provides a mask. It will also use quite a lot of memory, potentially negating the benefits from using memory-efficient kernels

File: torch/nn/functional.py [hunk 2]
```diff
@@ -4922,6 +4922,7 @@ def multi_head_attention_forward(
     key_padding_mask: Optional[Tensor] = None,
     need_weights: bool = True,
     attn_mask: Optional[Tensor] = None,
+    _causal_mask: bool = False
```
Comment (albanD) [inline torch/nn/functional.py:4925]: Just mentioning that doesn't really make the arguments private. It will still show up in the signature of the function on the doc page.
This function is not actually documented on the website so the point is moot-ish but you can see where it is mentioned in https://docs-preview.pytorch.org/90508/jit_builtin_functions.html#supported-pytorch-functions

Comment (mikekgfb) [reply torch/nn/functional.py:4925]: TBH, I don't think it should be private -- mirroring the concerns here from dan3, I think it will be preferable to give users the explicit option.

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:295]: I think the preferred way will be to surface this as the causal mask flag as part of the API.  I am/was hoping that we might rely on torch.compile() to constant fold it.  (And maybe, if it is unable to do so, we might have it optimized away by torch.compile() ?

File: torch/nn/modules/transformer.py [hunk 3]
```diff
@@ -278,8 +283,26 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
                 output = torch._nested_tensor_from_mask(output, src_key_padding_mask.logical_not(), mask_check=False)
                 src_key_padding_mask_for_layers = None
 
+        # Prevent type refinement
+        is_causal = False
+        if mask is not None:
+            if is_causal:
+                raise RuntimeError("specify either mask or is_causal, but not both")
+
+            # sz = mask.size(0)
```
Comment (albanD) [inline torch/nn/modules/transformer.py:292]: Should this be removed?

**Review by albanD — COMMENTED** [review]
Change sounds ok. Docstrings need to be updated.

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:292]: removed.

Comment (mikekgfb) [reply torch/nn/modules/transformer.py:292]: > Change sounds ok. Docstrings need to be updated.

Docstring updated.

File: torch/nn/modules/transformer.py [hunk 4]
```diff
@@ -623,8 +643,16 @@ def __setstate__(self, state):
...
-                tgt_key_padding_mask: Optional[Tensor] = None, memory_key_padding_mask: Optional[Tensor] = None) -> Tensor:
+    def forward(
+        self,
+        tgt: Tensor,
+        memory: Tensor,
+        tgt_mask: Optional[Tensor] = None,
+        memory_mask: Optional[Tensor] = None,
+        tgt_key_padding_mask: Optional[Tensor] = None,
+        memory_key_padding_mask: Optional[Tensor] = None,
+        is_causal: bool = False,
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/modules/transformer.py]: added

File: torch/nn/functional.py [hunk 5]
```diff
@@ -4929,6 +4929,7 @@ def multi_head_attention_forward(
     static_k: Optional[Tensor] = None,
     static_v: Optional[Tensor] = None,
     average_attn_weights: bool = True,
+    is_causal: bool = False,
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/functional.py:4932]: added

**Review by albanD — APPROVED** [review]
The change itself sounds ok.
You most likely want to rebase above master to avoid trunk failures.
